### PR TITLE
Fixes issues with concurrent write

### PR DESCRIPTION
**Pull request by nivemaham**
- Using explicit locking to prevent concurrent writing. This finally prevents multiple tasks processing the same record. 
- Adds specific exceptions to handle

### CODE REVIEW

File: kafka-connect-upload-source/src/main/java/org/radarbase/connect/upload/UploadSourceTask.kt [hunk 1]
```diff
@@ -15,6 +15,7 @@ import org.radarbase.connect.upload.util.VersionUtil
 import org.slf4j.LoggerFactory
 
 class UploadSourceTask : SourceTask() {
+    private var pollRecordSize = 1
```
Comment (blootsvoets) [inline kafka-connect-upload-source/src/main/java/org/radarbase/connect/upload/UploadSourceTask.kt]: `pollRecordSize` is not used yet

File: kafka-connect-upload-source/src/main/java/org/radarbase/connect/upload/api/UploadBackendClient.kt [hunk 2]
```diff
@@ -94,6 +96,8 @@ class UploadBackendClient(
             when (response.code()) {
                 401 -> throw NotAuthorizedException("access token is not provided or is invalid : ${response.message()}")
                 403 -> throw NotAuthorizedException("access token is not authorized to perform this request")
+                400 -> throw StaleStateException("Could not perform request due stale state: ${response.message()}")
```
Comment (blootsvoets) [inline kafka-connect-upload-source/src/main/java/org/radarbase/connect/upload/api/UploadBackendClient.kt]: I'm not sure if 400 is the right code for a stale state, the request is well-formed, the server just can't process it. Perhaps 425 Too Early or even just 409 Conflict may be better candidates.

File: radar-upload-backend/src/main/java/org/radarbase/upload/inject/DoaEntityManagerFactory.kt [hunk 3]
```diff
@@ -40,8 +33,10 @@ fun <T> EntityManager.transact(transaction: EntityManager.() -> T): T {
         try {
             getTransaction().commit()
         } catch (exe: Exception) {
-            logger.warn("Rolling back operation: {}", exe.toString())
-            getTransaction().rollback()
+            logger.error("Rolling back operation: {}", exe)
+            if(getTransaction() != null && getTransaction().isActive) {
```
Comment (blootsvoets) [inline radar-upload-backend/src/main/java/org/radarbase/upload/inject/DoaEntityManagerFactory.kt]: Would it make sense to use a single `val transaction = getTransaction()` variable and use that throughout this code block? The null-check would need to happen only once, and here we can just check whether the transaction is active.

File: kafka-connect-upload-source/src/main/java/org/radarbase/connect/upload/api/UploadBackendClient.kt (hunk 2 above)
Comment (nivemaham) [reply kafka-connect-upload-source/src/main/java/org/radarbase/connect/upload/api/UploadBackendClient.kt]: Then I would go for the 409. In this case, request is wellformed though it has an old state in the context on the server.

File: radar-upload-backend/src/main/java/org/radarbase/upload/inject/DoaEntityManagerFactory.kt (hunk 3 above)
Comment (nivemaham) [reply radar-upload-backend/src/main/java/org/radarbase/upload/inject/DoaEntityManagerFactory.kt]: Yep. Agreed

File: kafka-connect-upload-source/src/main/java/org/radarbase/connect/upload/UploadSourceTask.kt (hunk 1 above)
Comment (nivemaham) [reply kafka-connect-upload-source/src/main/java/org/radarbase/connect/upload/UploadSourceTask.kt]: My bad. Didn't mean to add it to the commit

**Pull request merged**